Git workflow

okonet opened the pull request:
This PR implements a custom git workflow that works as follows:

1. Before running linters, check if there are partially stated files (i.e. both with changes in commit index and the working copy)
1. Stash and remove from the working copy all files that aren't in the commit index
1. Execute commands (linters and formatters) on these files
1. Add formatters changes to index with `git add` (should be in the task config)
1. Restore the stashed working copy state if all linters exit with zero code (i.e. success)
1. Compare new commit index and working copy and apply as many formatting changes to working copy as possible. If there are conflicting hunks, drop formatter's changes. User modifications that were stashed should always take precedence over formatters.

Fixes #62

### Related issues / tasks to complete

- [x] Handle Ctrl+C before stashing (#386)
- [x] Should not restore deleted files (#387)
- [x] Only stash/unstash if partially staged files exist. Now it only checks if unstaged files exist.
- [x] Handle appearance of `.rej.xx` files. Probably just deletion of them is okay since we don't care about conflicts and always restore user's state.

okonet committed:
Use `git stash --keep-index` to enable linting and committing only staged changes! Closes #62

okonet committed:
Added gitWorkflow and basic tests

okonet committed:
More work on tests + some refactoring

okonet committed:
Added more tests + fixed some bugs

okonet committed:
Work with absolute and relative paths. Added tests.

okonet committed:
Switched to gitWorkflow

okonet committed:
Added test user name and email to try to fix Travis

okonet committed:
Bumped version to 4.0.0-beta-1

okonet committed:
Bumped version

okonet committed:
Extracted expect extension to a separate file. Added a failing test for gitWorkflow.

Use a .toEventuallyEqual() for git status checks

okonet committed:
Switched to async/await for tests

okonet committed:
Removed no-console eslint error

okonet committed:
Some progress on git workflow.

- Switched to ghooks instead of pre-commit
- Added pre and post scripts that run `git stash` and `git stash pop`

okonet committed:
Merge branch 'master' into git-worflow

okonet committed:
tests: Migrate git workflow tests to Jest

okonet committed:
chore: Switch to husky and fix dependencies

okonet committed:
chore: Add *.json to .lintstagedrc config

okonet committed:
tests: Remove dependency to expect

okonet committed:
tests: Update tests

okonet committed:
Use gitRestore in post command

okonet committed:
tests: Update tests and add gitRestore

okonet committed:
test: Skip failing test for now

okonet committed:
Trying a proposed approach and restoring from stash before the commit

See https://github.com/okonet/lint-staged/issues/62#issuecomment-286830310

okonet committed:
tests: Removed unnecessary code, test, and snapshot

koddsson commented:
Hey @okonet. Is this branch safe to test out? ie. are there any glaring issues where I might lose data if I wanna check out this branch and maybe help test since I'm really excited about this feature 😊

okonet commented:
Unfortunately, it is still a WIP but you can try it out and report issues. I would not use on your project's code though since it still may loose some data AFAIK.

koddsson commented:
Ok I'll see if I can use it some side projects rather than code at work 💃

Stupidism commented:
@okonet , more detailed description: https://github.com/okonet/lint-staged/issues/62#issuecomment-299525019
## Solution with npm scripts and bash scripts
```
  "scripts": {
    "stash-unstaged": "git stash save -k --include-untracked 'unstaged-stash' >> /dev/null",
    "pop-if-gst-clean": "./tools/pop-if-gst-clean.sh",
    "lint-staged": "lint-staged || (npm run pop-stash >> /dev/null && exit 1)",
    "pop-if-gd-clean": "./tools/pop-if-gd-clean.sh",
    "stash-linted": "git stash save 'linted-stash' >> /dev/null",
    "stash-pop-unstaged": "git stash pop stash@{1}",
    "stash-pop-linted": "git read-tree stash && git stash drop",
    "clean-lint-staged": "lint-staged >> /dev/null"
  },
  "//pre-commit": {
    "//reference-links": {
      "github": "https://github.com/okonet/lint-staged/issues/62#issuecomment-286830310",
      "stackoverflow": "http://stackoverflow.com/questions/13889242/26685296#26685296"
    },
    "//stash-pop-linted": "use `git read-tree` to resolve conflicts",
    "//clean-lint-staged": [
      "Re-run `lint-staged` for fixable style errors mistakenly recovered to be unstaged by `read-tree`",
      "NOTICE: Although they are fixed, you will see them when input a commit-message",
      "Side effect is it will auto-fix unstaged part in these staged files",
      "Feel free to remove this cmd if you don't want the side effect or just feel adventurous",
      "If so, don't be surprised when you see unstaged&fixable style errors after commit"
    ]
  },
  "pre-commit": [
    "stash-unstaged",
    "pop-if-gst-clean",
    "lint-staged",
    "pop-if-gd-clean",
    "stash-linted",
    "stash-pop-unstaged",
    "stash-pop-linted",
    "clean-lint-staged"
  ]
```

Commit:
Merge branch 'master' into git-worflow

Commit:
refactor: Fix ESLint errors, run prettier, remove yarn.lock

Commit:
refactor: Use functions and normal modules.export instead of class

Commit:
feat: Add git stash and restore workflow

Commit:
Move git workflow to runAll and use Listr for running it

Commit:
Only stash/unstash files if there are unstaged files found

Commit:
Use base options for root listr tasks

Commit:
Better tests for git workflow

Commit:
WIP on better partially staged patch application and conflict resolution

Commit:
Reduce boilerplate in test

Commit:
Add a failing test for partially staged files

Commit:
refactor: Remove unused jest-stdout-serializer

Commit:
Create more git utilities for staged / unsteaged files

Commit:
tests: Add a failing test for deletion

Commit:
tests: More tests for git workflow

Commit:
Fix execution order problem

Commit:
Run in verbose mode

Commit:
tests: Update test case and comment the conent check for now

Commit:
Generate path if there are unstaged files

Commit:
tests: Skip the failing test

Commit:
tests: Update snapshots

Commit:
Merge branch 'master' into git-worflow

Commit:
chore: Use debug instead of console.log

Commit:
Run lint-staged in debug mode

Commit:
Remove unused code that fails tests on Node 4

Commit:
fixup! chore: Use debug instead of console.log

okonet commented:
@sudo-suhas windows fail are failing and I don't know how to fix them. Any help?

sudo-suhas commented:
I got similar errors when I ran the tests on my local system. So it's not something specific to appveyor. I will need to investigate this to be able to understand the reason. I am slightly caught up with something at work. Please give me some time for me to come back to this.

okonet commented:
I want to push this today since I'm presenting at Agent Conf and I'd love to ship it. I will create a separate issue for tests on Windows. Thanks!

Commit:
Remove unused post.js and pre.js files

sudo-suhas commented:
Please do share the link for the talk when you can. I'd love to watch it.

okonet commented:
I'm gonna do that! Thanks!

Commit:
Add stashUnstaged flag to the config

luftywiranda13 commented:
@okonet @sudo-suhas consider to take a look at [g-status](https://github.com/luftywiranda13/g-status), it'll help us reduce many codes in this PR.

we can use git status codes to filter out some files not just in the staging area but also in the working tree.  (vs `staged-git-files` which ignores the condition in the working tree). so we know which files are fully staged or partially staged

but yeah, it requires node >= 6

okonet commented:
That’s amazing! Thanks!

luftywiranda13 commented:
@okonet  I made it as a tool for my simple use case. if you find it not flexible enough, maybe I can give both you and @sudo-suhas access to the repo

Commit:
Fix tests

Commit:
Update test title

sudo-suhas commented:
Hey @okonet, I have investigated the issue with git on Windows. I am still not entirely sure why this would happen only on Windows. It has to do with the way the git repo is initialised which results in a [`bare` repo](https://git-scm.com/docs/git-init#git-init---bare):

```
C:\Users\SUHAS~1.KAR\AppData\Local\Temp\tmp-178167ccx8x3YbevL
λ ls
test.css  test.js

C:\Users\SUHAS~1.KAR\AppData\Local\Temp
λ git --git-dir C:\Users\suhas.karanth\AppData\Local\Temp\tmp-178167ccx8x3YbevL\.git init
Initialized empty Git repository in C:/Users/suhas.karanth/AppData/Local/Temp/tmp-178167ccx8x3YbevL/.git/

C:\Users\SUHAS~1.KAR\AppData\Local\Temp\tmp-178167ccx8x3YbevL (master -> origin)
λ git status
fatal: This operation must be run in a work tree

C:\Users\SUHAS~1.KAR\AppData\Local\Temp\tmp-178167ccx8x3YbevL (master -> origin)
λ cat .\.git\config
[core]
        repositoryformatversion = 0
        filemode = false
        bare = true
        symlinks = false
        ignorecase = true

C:\Users\SUHAS~1.KAR\AppData\Local\Temp\tmp-178167ccx8x3YbevL (master -> origin)
λ rm -rf .git

C:\Users\SUHAS~1.KAR\AppData\Local\Temp
λ git init C:\Users\suhas.karanth\AppData\Local\Temp\tmp-178167ccx8x3YbevL
Initialized empty Git repository in C:/Users/suhas.karanth/AppData/Local/Temp/tmp-178167ccx8x3YbevL/.git/

C:\Users\SUHAS~1.KAR\AppData\Local\Temp\tmp-178167ccx8x3YbevL (master -> origin)
λ git status
On branch master

No commits yet

Untracked files:
  (use "git add <file>..." to include in what will be committed)

        test.css
        test.js

nothing added to commit but untracked files present (use "git add" to track)

C:\Users\SUHAS~1.KAR\AppData\Local\Temp\tmp-178167ccx8x3YbevL (master -> origin)
λ cat .\.git\config
[core]
        repositoryformatversion = 0
        filemode = false
        bare = false
        logallrefupdates = true
        symlinks = false
        ignorecase = true
```

@luftywiranda13, I saw that you had linked to [`simple-git`](https://github.com/steveukx/git-js) from `g-status`. Could you please help me understand what is the use-case which `g-status` solves(which `simple-git` does not)?

@okonet I tried using `simple-git` to do the init and it worked correctly:

```js
'use strict';

const gitP = require('simple-git/promise');

const git = gitP('C:\\Users\\suhas.karanth\\AppData\\Local\\Temp\\tmp-178167ccx8x3YbevL');

async function main() {
    try {
        await git.status()
    } catch(err) {
        // Error is logged to stderr by default
        // console.log(err.message)
    }

    try {
        await git.init()
        console.log(await git.status())
    } catch(err) {
        console.log(err.message)
    }
}

main()
```

```
λ node test-git.js
fatal: Not a git repository (or any of the parent directories): .git

StatusSummary {
  not_added: [ 'test.css', 'test.js' ],
  conflicted: [],
  created: [],
  deleted: [],
  modified: [],
  renamed: [],
  files:
   [ { path: 'test.css', index: '?', working_dir: '?' },
     { path: 'test.js', index: '?', working_dir: '?' } ],
  ahead: 0,
  behind: 0,
  current: 'No',
  tracking: null }
```

I think it would be better to use a well tested module for interacting with `git`. This should reduce the amount of code needed as well.

sudo-suhas commented on package.json:
```diff
@@ -35,6 +35,7 @@
     "dedent": "^0.7.0",
     "execa": "^0.8.0",
     "find-parent-dir": "^0.3.0",
+    "fs-promise": "^2.0.3",
```
[`fs-promise` is deprecated](https://github.com/kevinbeaty/fs-promise). Moreover, we could just use `pify` to promisify `fs`:

```js
λ node
> const pify = require('pify')
undefined
> const fs = pify(require('fs'))
undefined
> fs.readFile('./.lintstagedrc', 'utf8').then(console.log)
> {
    "linters": {
        "*.js": [
            "eslint --fix",
            "git add"
        ],
        ".*{rc, json}": "jsonlint"
    }
}
```

okonet commented:
@sudo-suhas appreciate you doing this! I will take a look at simple-git and most possibly switch to it. I could not find anything good when I started the branch almost 1 year ago I think.

sudo-suhas commented on test/gitStash.spec.js:
```diff
@@ -0,0 +1,430 @@
+    // Create JS file
+    await fsp.writeFile(
+      path.join(wcDirPath, 'test.js'),
+      `module.exports = {
```
Can we use `dedent` for these multi-line strings? Note that it would trim any leading or trailing white space including `\n`.

sudo-suhas reviewed (changes requested):
I have not taken a through look at the logic for `runAll` with unstaged files and the tests. Please let me know if you'd like me to do that.

sudo-suhas commented on test/gitStash.spec.js:
```diff
@@ -0,0 +1,430 @@
+      expect(await gitStashList()).toEqual('')
+    })
+
+    it('should add hooks fixes to index and working copy on partially staged files', async () => {
```
@okonet, could you fix this test? It is failing on CI - https://travis-ci.org/okonet/lint-staged/jobs/334488727#L507

luftywiranda13 replied:
I think to write tests in async/await, the correct format is `await expect(fn()).resolves.toEqual(...)`

luftywiranda13 replied:
Or is it just a convention?

SimenB replied:
You can use `resolves` if you want, but just using `await` is often cleaner. Depends on your asserition.

```js
await expect(someFunc()).resolves.toBe('cool');
```

vs
```js
const value = await someFunc();
expect(value).toBe('cool');
```

luftywiranda13 replied:
@SimenB ah I see, thanks! 👍🏻

luftywiranda13 commented:
@sudo-suhas basically https://github.com/luftywiranda13/g-status/blob/master/readme.md#why concludes it. 

I need to filter out the results of `git status` command to be used in two of my internal projects. 

So I made g-status in order to make it **reusable** and make me don’t have to reimplement the same logic and unit tests over and over. 

It is possible to use `.status()` from `simple-git` directly but it’s not that **simple** to be honest, and also simple-git doesn’t provide patterns based matching and filters

Siyfion commented:
Wowsers, I had no idea this was going to be such a massive headache-issue! Still, it looks like there's been some really decent progress, what's the status of this at the moment? It's working with Linux and macOS, but not Windows (is that the gist of it?).

Also, you guys are now responsible for the **1** recommended integration strategy with Prettier.io (https://prettier.io/docs/en/precommit.html)

Commit:
Check if there are staged files before stashing

Commit:
Add logging for git commands

Commit:
chore: Use Node >=8 to supoort async / await

BREAKING CHANGE

Node >= 8.x is required

Commit:
chore: Disable prettier eslint plugin in .eslintrc

By disabling it in .eslintrc and enabling as a `--rule` option
on npm script allows removing formatting errors in editors
but helps keep things formatted with lint-staged and fail on CI

Commit:
chore: Use prettier for *.js, *.json and *.md files

Commit:
test: Better test for partially-staged files

Commit:
Use task context and set hasErrors if linters fail

Commit:
Cleaner test name

Commit:
Add update index step to runAll which should only be executed if all linters pass

Commit:
chore: Update yarn.lock

Commit:
WIP on better gitWorkflow

Commit:
Use async/await for patch strategy

Commit:
Better test names

Commit:
fix test

Commit:
Upgrade Jest

Commit:
Use async/await for git workflow. Return stdout from execGit

Commit:
Remove console.log

Commit:
fixup! Use async/await for git workflow. Return stdout from execGit

okonet commented:
It's not that simple. You can't `git stash pop` without conflicts if you've modified index. Please take a look at the implementation.

NoChapter commented:
@okonet  Please see https://github.com/okonet/lint-staged/issues/62#issuecomment-411964870, `git stash pop` just works. 
Sure, We should take care the conflicts ourselves after popped.

okonet commented:
> git stash pop just works.

I have a different experience with it.

Commit:
tests: Fix tests for gitWorkflow.

Add new lines to partially-staged test.

Commit:
Implement the thing!

Commit:
Remove patch file after applying it. Add debug message

Commit:
When formatted changes could not be applied, we still want to format index

Commit:
fixup! tests: Fix tests for gitWorkflow.

Commit:
fix

Commit:
fix(git): Fix hasUnstagedFiles check + add test

Commit:
Make the partially staged test more realistic

Commit:
Fix the partially staged case by adding a EOF to patch file

Commit:
Merge branch 'master' into git-worflow

Commit:
v8.0.0-beta.0

Commit:
fix: only run stash/unstash sequence when partially staged files detected

Commit:
fix: Add cleanup function and use named variables for refs

Commit:
refactor: Remove unneded tree check

Commit:
fix: Restore working copy when terminated

Fixes #386

okonet commented:
@sudo-suhas @luftywiranda13 this is shaping out well. Do you mind testing and reviewing it? At this point I would consider it done but I'd like to clean up code a bit more. Also, it feels like before we restore working copy we need a more robust way of detecting the need to do so. Thoughts?

okonet commented:
@sudo-suhas should we remove Node v6 support as a separate PR so tests for this one won't be affected? I'd like to release this as v8.0 though...

Commit:
refactor: remove unused `stashUnstaged` option

Commit:
fix: check if `git status` returns a string

Commit:
v8.0.0-beta.1

Commit:
test: mock gitWorkflow module in runAll tests

trusktr commented:
(continuing from https://github.com/okonet/lint-staged/issues/62#issuecomment-420022062)

> Not sure what you trying to achieve but you didn’t try to commit partially staged files so it worked as expected

Yes, if you see my output there:

```
❯ git status
On branch use-jss
Your branch and 'origin/use-jss' have diverged,
and have 3 and 1 different commits each, respectively.
  (use "git pull" to merge the remote branch into yours)

Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

	modified:   src/components/Footer.tsx
	modified:   src/components/Header.tsx
	modified:   src/components/HeaderMobileNavModal.tsx
	modified:   src/components/RootLayout.tsx

Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   package.json
	deleted:    src/App.vue
	deleted:    src/names.js
	modified:   src/pages/_document.tsx
	modified:   src/styles/theme.ts

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	src/App.vuee src/namess.js
```

The `Changes to be committed` are my staged changes (whole files in this case, I haven't tried partially staged files).

There are `deleted` unstaged changes there too, which are somehow `reset` by lint-staged after the commit completes.

Let's test this by combinations of the following:

- have unstaged deleted files
- unstaged renamed files (has both deleted and untracked)
- untracked files (are not a rename)
- modified files

and with those changes, do one of each

- add all changes
- add some changes
- add only changes of one type
- etc

This is how I'm finding that it isn't working like I've described in the other thread.

trusktr commented:
In the simple case where I have only `modified` changes, this seems to work fine so far. Just try the other cases above.

okonet commented:
In the described above case it won’t do anything different from before so I’m not sure why you experiencing it. Please make sure you’re at beta1 but I think you are. 

Also I will try to add your cases as test cases but feel free to jump on and modify tests yourself so we aren’t playing broken phone game.

Commit:
fix: Do not display skip message for stashing step if stashed

Since it was missing the explicit return after stashing, it was marking the task as skipped.

Fixes #490

Commit:
tests: Add more tests for stashing and restoring

Commit:
tests: Mock gitWorkflow in index.spec. Test non-zero exitCode

sudo-suhas commented on test/gitStash.spec.js:
```diff
@@ -0,0 +1,477 @@
+      expect(res).toMatchSnapshot()
+    })
+
+    it('should return an array of partially staged files', async () => {
```
I don't think this test is written correctly. The snapshot output:
```js
exports[`git getFilesStatus should return an array of partially staged files 1`] = `
Array [
  "M  test.css",
  "MM test.js",
]
`;
```

okonet replied:
Removed and replaced code with https://github.com/luftywiranda13/g-status

sudo-suhas commented on test/gitWorkflow.spec.js:
```diff
@@ -0,0 +1,76 @@
+  })
+
+  describe('execGit', () => {
+    beforeEach(() => {
```
`beforeEach` can be removed?

sudo-suhas commented on test/gitWorkflow.spec.js:
```diff
@@ -0,0 +1,76 @@
+
+    it('should return result from stdout', async () => {
+      const res = await gitflow.execGit(['branch', '--list'])
+      expect(res).toMatchSnapshot()
```
Is this working as intended? Snapshot output:
```js
exports[`gitWorkflow execGit should return result from stdout 1`] = `"output"`;
```

okonet replied:
Removed since it was mocked

sudo-suhas commented on test/__mocks__/execa.js:
```diff
@@ -1,9 +0,0 @@
-module.exports = jest.fn(() =>
```
This seems to be repeated in 3 different files. Could you please tell me what was the reason for removing it?

okonet replied:
Some tests relying on unmocked `execa` version which I could not find how to override in the test. Any hints?

sudo-suhas replied:
`jest.dontMock('execa')`?

sudo-suhas reviewed (dismissed):
> Also, it feels like before we restore working copy we need a more robust way of detecting the need to do so. Thoughts?

Sorry, can't help much here as I don't know enough about git and found it a little hard to comprehend the current solution.

> @sudo-suhas should we remove Node v6 support as a separate PR so tests for this one won't be affected? I'd like to release this as v8.0 though...

I don't think this is necessary. We are not really doing any changes for migration from 6 to 8 so segregating the changes doesn't come into the picture. Releasing these changes as v8 to indicate the breaking change seems good enough to me.

Also,
- Please remove node v6 from the test matrix on CI.
- We had discussed Windows related failures earlier this year. Please see https://github.com/okonet/lint-staged/pull/75#issuecomment-361527083 which went into details of what was responsible for the issues. Please address this.

Commit:
feat: Add whitespace between errors and make error messages stand out

Commit:
feat: Better task name and message for update stash step

Commit:
fix: use the forked version of listr-update-renderer until the fix is merged in

Commit:
fix: upgrade execa to v1.0.0

Commit:
refactor: Cleanup gitWorkflow module

Commit:
feat: detect SIGINT termination for linter tasks and exit appropriately

Commit:
fix: do not exit main Listr process on SIGINT

Fixes #386

Commit:
refactor: add a return from the main process

Commit:
v8.0.0-beta.2

trusktr commented:
I just ran into a commit leaving a `header.tsx.rej` file behind, containing

```
diff a/src/components/Header.tsx b/src/components/Header.tsx	(rejected hunks)
@@ -478,4 +483,6 @@ function styles(theme: Theme) {
-                  background: Color(navyBlue).darken(0.1).string(),
-                },
-              },
-            },
+                  background: Color(navyBlue)
+                    .darken(0.1)
+                    .string()
+                }
+              }
+            }
```

What does that mean? I haven't tried the latest beta yet.

okonet commented:
I hit this one once myself. This probably means that the hunk was reformatted although it was also edited but not selected for commit. During the restoration process it could not cleanly apply both edits so it put it into rej file. I’m not sure yet how to detect such cases. Can you provide me with exact actions you did so I could try creating a test case?

trusktr commented:
> Can you provide me with exact actions you did so I could try creating a test case?

I'm flying through commits at the moment, but I'll see if I can make a small reproduction the next time I encounter it.

On another note, I just tried beta.2, but I still see the confusing message in the output:

```
  ↓ Stashing changes... [skipped]
    → No unstaged files found...
```

I'm confused because my expectation is that it will tell me it found unstage changes to stash (in fact this is what it does and is working), but the message makes it seems like I have no unstaged changes to stash (which isn't what actually happens).

Commit:
tests: Move mock file to /test

okonet commented:
> I'm confused because my expectation is that it will tell me it found unstage changes to stash

I see why you're confused and I'll update the copy to match the behavior. Which is at the moment: if there are no partially staged files, do not perform stash / restore.

fabb commented:
Have you considered solving the requirement of linting/formatting only staged files without stashing? This tool might serve as a showcase: https://github.com/hallettj/git-format-staged/blob/master/git-format-staged

okonet commented:
There is #62 with related discussions. Let’s keep comments here related to the work in this PR please.

okonet commented on src/gitWorkflow.js:
```diff
@@ -0,0 +1,197 @@
+}
+
+async function generatePatchForTrees(tree1, tree2, options) {
+  // TODO: Use stdin instead of the file for patches
```
Yes, I still this is still relevant. I will need to spend more time investigating this, though. I'd like to release v8 and work on these things when I have more time.

okonet commented on src/gitWorkflow.js:
```diff
@@ -0,0 +1,197 @@
+  return []
+}
+
+async function execGit(cmd, options) {
```
I'll try switching to `simple-git`, thanks!

okonet replied:
@sudo-suhas after an hour I'm giving up. It doesn't look like the simple-git supports `--git-dir` attribute of `git` that is needed for tests to work. I could not figure out how to solve it. Are you still having issues on Windows? Can you or @luftywiranda13 help me with migrating it to simple-git then?

okonet replied:
Hmm, interestingly I've removed `gitDir` option and all tests are green. But if I'm using `simple-git` I'm getting some weird SHA errors. Please @luftywiranda13 can you help me with this?

sudo-suhas replied:
Did you take a look at the code example with `simple-git` in https://github.com/okonet/lint-staged/pull/75#issuecomment-361527083.

okonet replied:
Yes, I did. I tried implementing my code using simple-git and failed with some weird errors coming from git. I'm not sure the root cause yet so if you want you can investigate this after this is released.

okonet commented on src/gitWorkflow.js:
```diff
@@ -0,0 +1,197 @@
+const fsp = require('fs-promise')
+const debug = require('debug')('lint-staged:git')
+
+let workingCopyTree = null
```
What are benefits besides the complexity? There are no multiple instances of lint-staged that could run in parallel so I think that's fine for now. Ideally, the code would not rely on the state at all (that's my goal)

sudo-suhas replied:
Just a rule of thumb I follow. A `class` would be explicit for what's the state and what are the methods which access/modify it.

okonet replied:
As I said, I don't think the state is needed but I really want to release it as is and refactor after that since it's been 2.5 years in making.

Commit:
fixup! chore: Use Node >=8 to supoort async / await

Commit:
tests: Remove redundant test

Commit:
refactor(gitWorkflow): Use g-status and remove custom partially staged files detection

Commit:
refactor(gitWorkflow): Remove redundant gitDir option

okonet commented on test/gitStash.spec.js:
```diff
@@ -0,0 +1,477 @@
+      expect(res).toMatchSnapshot()
+    })
+
+    it('should return an empty array for only modified files', async () => {
```
Removed and replaced code with https://github.com/luftywiranda13/g-status

okonet commented:
@sudo-suhas latest Windows tests fail with some strange snapshot differences. Could you please try running locally? https://ci.appveyor.com/project/okonet/lint-staged/builds/19651280/job/71b9px4salpqsokj

Commit:
fix(gitWorflow): Make the copy for partially staged case more clear

okonet commented:
@trusktr I've resolved this in fa492abb931043a5ebf8dc2f93ff89a928c60d99

sudo-suhas commented:
I am able to see the same behavior on my Windows system.  The error message is coming from here - https://github.com/okonet/lint-staged/blob/fa492abb931043a5ebf8dc2f93ff89a928c60d99/src/resolveTaskFn.js#L63-L65

okonet commented:
Hmm, I’m not sure how to test it then. Ideas?

sudo-suhas commented:
Could we reduce the scope of changes introduced? I know that's easier said than done but in all fairness, this PR does a lot of changes.

okonet commented:
It does only required changes. There is no way we can reduce the scope I’m afraid.

Commit:
tests(windows): Use mock of log-symbols to match snapshots on Mac/Windows

Commit:
Merge branch 'snapshots-win' into git-worflow

Commit:
tests: Update snapshots

Commit:
tests: Replace CLRF with LF in tests to normalize on Windows

Commit:
fix: Fix fail message when termination signal is detected

Commit:
fix: Fix the case where failure message could reach impossible state

Commit:
fixup! chore: Use Node >=8 to supoort async / await

okonet commented on src/gitWorkflow.js:
```diff
@@ -0,0 +1,197 @@
+    })
+    return stdout
+  } catch (err) {
+    throw new Error(err)
```
Not sure what you mean. What wrapping? Without it will be unhandled promise case.

trusktr replied:
What happens in the unhandled case?

I believe newer versions of v8 (therefore Node) handle unhandled promise rejections as regular errors now, if that's what you were trying to guard against. Maybe it's good to play it safe, just in case, for people on older versions of Node.

okonet commented on src/gitWorkflow.js:
```diff
@@ -0,0 +1,197 @@
+  // Save ref to the current index
+  indexTree = await writeTree(options)
+  // Add working copy changes to index
+  await execGit(['add', '.'], options)
```
Yes, we want to restore the state after we're done so we need to add everything before we clean up and get the reference to the treeish (next line). Anyways there is a test for this case which restores deleted files as "Deleted" (where `git stash` doesn't)

okonet commented on src/gitWorkflow.js:
```diff
@@ -0,0 +1,197 @@
+  formattedIndexTree = await writeTree(options)
+}
+
+async function applyPathFor(tree1, tree2, options) {
```
Good catch!

okonet commented on src/runAll.js:
```diff
@@ -56,13 +58,57 @@ module.exports = function runAll(config) {
       }
     }))
 
+    const listrBaseOptions = {
```
It should be used for all instances, that's why I extracted it. Or what exactly do you mean?

okonet commented on src/runAll.js:
```diff
@@ -56,13 +58,57 @@ module.exports = function runAll(config) {
+        [
+          {
+            title: 'Stashing changes...',
+            enabled: () => filenames.length > 0,
```
Yeah, probably it could. I'll take a closer look.

okonet commented on src/runAll.js:
```diff
@@ -56,13 +58,57 @@ module.exports = function runAll(config) {
+            task: () =>
+              new Listr(
+                tasks,
+                Object.assign({}, listrBaseOptions, {
```
It's not supported natively AFAIK. Or it could just be old code here. Not sure.

okonet replied:
Yep, no spread in Node 8

SimenB replied:
It's in node 8.6 and newer: https://node.green/#ES2018-features-object-rest-spread-properties-object-spread-properties

okonet replied:
Thanks, I've updated the minimum version to be 8.6

okonet commented:
Finally all tests are green on both linux and Windows!

Commit:
refactor: Fix typo

Commit:
refactor: Use `skip` instead of `enabled` for stashing task

Commit:
chore: Remove fs-promise in favor of pify and native fs

Commit:
chore: Do not use debug flag for lint-staged

Commit:
tests: Restore execa mock and remove redundant code

Commit:
tests: Remove unnecessary stash checks since we don't rely on git stash

okonet reviewed (commented):
I've made requested changes and I think this is good for the final review.

okonet commented:
@sudo-suhas can you please test the beta version on Windows? I believe your comment regarding git ops is concerting with tests and I'm able to run tests on CI and my local windows machine. Although I'm struggling the CRLF issues on it so I couldn't yet test the package in real life scenario (BTW what's the setup you're using to deal with CRLF issue?)

Commit:
Merge remote-tracking branch 'origin/master' into git-worflow

Commit:
tests: Switch some tests to inline snapshots

Commit:
tests: Add more tests for renamed and untracked files

okonet commented:
@trusktr I've added a few tests cases for handling of renamed and untracked files in https://github.com/okonet/lint-staged/pull/75/commits/da2adf9a066749bcea74910a86d9de9791f602e5. 

Can you please take a look and let me know if this is what you'd expect? It looks right to me.

Commit:
v8.0.0-beta.3

Commit:
refactor: Use stdin for patch instead of writing a the file

Commit:
fixup! chore: Use Node >=8 to supoort async / await

Commit:
chore: Remove unused dependency

Commit:
fix: Delete *.rej files

Commit:
fix(git): Always add a newline to the end of the patch.

This fixes a case when the patch would fail with

`error: fatal: corrupt parch at line XX`

caused by git

Commit:
v8.0.0-beta.4

Commit:
fix(git): Fix hasPartiallyStagedFiles return true for untracked files

okonet commented on src/gitWorkflow.js:
```diff
@@ -0,0 +1,197 @@
+  cleanup()
+}
+
+async function gitStashSave(options) {
```
No, that's a good catch. It was due to experimenting with different approaches.

Commit:
chore: Add a TODO

Commit:
refactor: Use Object spread

Commit:
test: Add jest-snapshot-serializer-ansi

Commit:
refactor: Remove impossible case handling from makeErr

Commit:
refactor: Rename function to match exports

Commit:
chore: Revert version in package.json

koggdal commented on package.json:
```diff
@@ -30,12 +30,16 @@
-    "listr": "^0.14.1",
+    "listr": "^0.14.2",
+    "listr-update-renderer":
+      "https://github.com/okonet/listr-update-renderer/tarball/upgrade-log-update",
```
Hi, would it be possible to get this released as a version on npm instead of depending on it through a loose git branch?

eaviles replied:
Having issues to install anything that depends on lint-staged now. We are behind a corporate proxy here.

joeljeske replied:
Please use a npm published version!
